hal/aurix_tc3xx.c

```diff
@@ -0,0 +1,511 @@
+        type = IfxFlash_FlashType_D0;
+    }
+    else if (addr >= IFXFLASH_PFLASH_P0_START && addr <= IFXFLASH_PFLASH_P0_END)
+    {
+        type = IfxFlash_FlashType_P0;
+    }
+    else if (addr >= IFXFLASH_PFLASH_P1_START && addr <= IFXFLASH_PFLASH_P1_END)
+    {
+        type = IfxFlash_FlashType_P1;
+    }
+    else {
+        /* bad address, panic for now */
+        wolfBoot_panic();
+    }
+
+    return type;
+}
+
+/* Programs a single page in flash */
+static void RAMFUNCTION programPage(uint32_t address,
+                                    const uint32_t* data,
+                                    IfxFlash_FlashType type)
+{
+    if (address % IFXFLASH_PFLASH_PAGE_LENGTH != 0) {
+        wolfBoot_panic();
+    }
+
+    const uint16 endInitSafetyPassword =
+        IfxScuWdt_getSafetyWatchdogPasswordInline();
+
+    IfxFlash_enterPageMode(address);
+    IfxFlash_waitUnbusy(FLASH_MODULE, type);
+
+    for (size_t offset = 0;
+         offset < IFXFLASH_PFLASH_PAGE_LENGTH / sizeof(uint32_t);
+         offset += 2)
+    {
+        IfxFlash_loadPage2X32(address, data[offset], data[offset + 1]);
+    }
+
+    IfxScuWdt_clearSafetyEndinitInline(endInitSafetyPassword);
+    IfxFlash_writePage(address);
+    IfxScuWdt_setSafetyEndinitInline(endInitSafetyPassword);
+
+    IfxFlash_waitUnbusy(FLASH_MODULE, type);
+}
+
+/* Performs a hardware erase verify check on the range specified by address and
+ * len. Returns true if the region is erased */
+static int RAMFUNCTION flashIsErased(uint32_t address,
+                                     int len,
+                                     IfxFlash_FlashType type)
+{
+    uint32_t base = 0;
+
+    IfxFlash_clearStatus(UNUSED_PARAMETER);
+
+    /* sector granularity */
+    if (len > IFXFLASH_PFLASH_WORDLINE_LENGTH) {
+        base = GET_SECTOR_ADDR(address);
+        IfxFlash_eraseVerifySector(base);
+        IfxFlash_waitUnbusy(FLASH_MODULE, type);
+    }
+    /* wordline granularity */
+    else if (len > IFXFLASH_PFLASH_PAGE_LENGTH) {
+        base = GET_WORDLINE_ADDR(address);
+        IfxFlash_verifyErasedWordLine(base);
+        IfxFlash_waitUnbusy(FLASH_MODULE, type);
+    }
+    /* page granularity */
+    else if (len > 0) {
+        base = GET_PAGE_ADDR(address);
+        IfxFlash_verifyErasedPage(base);
+        IfxFlash_waitUnbusy(FLASH_MODULE, type);
+    }
+    /* error on 0 len for now */
+    else {
+        wolfBoot_panic();
+    }
+
+    /* No erase verify error means block is erased */
+    return (DMU_HF_ERRSR.B.EVER == 0) ? 1 : 0;
+}
+
+/* Returns true if any of the pages spanned by address and len are erased */
+static int RAMFUNCTION containsErasedPage(uint32_t address,
+                                          size_t len,
+                                          IfxFlash_FlashType type)
+{
+    const uint32_t startPage = GET_PAGE_ADDR(address);
+    const uint32_t endPage   = GET_PAGE_ADDR(address + len - 1);
+
+    for (uint32_t page = startPage; page <= endPage;
```

Comment:
Mixed declaration. Move variable definitions to top. Several other places

Reply:
done